File: golang/tests/pkg/integration/session_pause_resume_integration_test.go
package integration

import (
	"fmt"
	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay"
	"os"
	"testing"
	"time"
)

// TestSessionPauseResumeIntegration tests successful pause and resume operations on a session
func TestSessionPauseResumeIntegration(t *testing.T) {
	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Fatal("AGENTBAY_API_KEY environment variable is not set")
	}

	client, err := agentbay.NewAgentBay(apiKey, nil)
	if err != nil {
		t.Fatalf("Failed to initialize AgentBay client: %v", err)
	}

	fmt.Println("Creating a new session for pause/resume testing...")
	createResult, err := client.Create(nil)
	if err != nil {
		t.Fatalf("Failed to create session: %v", err)
	}
	createdSession := createResult.Session
	sessionId := createdSession.SessionID
	t.Logf("Session created with ID: %s", sessionId)

	// Ensure cleanup
	defer func() {
		fmt.Println("Cleaning up: Deleting the session...")
		deleteResult, err := createdSession.Delete()
		if err != nil {
			t.Logf("Failed to delete session: %v", err)
		} else {
			t.Logf("Session %s deleted successfully", sessionId)
			if !deleteResult.Success {
				t.Logf("Session deletion failed")
			}
		}
	}()

	// Verify session is initially in RUNNING state
	fmt.Println("Step 1: Verifying session is initially RUNNING...")
	getResult, err := client.GetSession(sessionId)
	if err != nil {
		t.Fatalf("Failed to get session: %v", err)
	}
	if !getResult.Success {
		t.Fatalf("Get session failed: %s", getResult.ErrorMessage)
	}
	if getResult.Data == nil {
		t.Fatal("Get session returned nil data")
	}
	if getResult.Data.Status != "RUNNING" {
		t.Fatalf("Expected session status RUNNING, got %s", getResult.Data.Status)
	}
	t.Logf("✓ Session is RUNNING: %s", sessionId)

	// Pause the session
	fmt.Println("Step 2: Pausing session...")
	pauseResult, err := client.Pause(createdSession, 600, 2.0)
	if err != nil {
		t.Fatalf("Failed to pause session: %v", err)
	}
	if pauseResult == nil {
		t.Fatal("Pause returned nil result")
	}
	if !pauseResult.Success {
		t.Fatalf("Pause failed: %s", pauseResult.ErrorMessage)
	}
	t.Logf("✓ Session pause initiated successfully")
	t.Logf("  Request ID: %s", pauseResult.RequestID)

	// Wait a bit for pause to complete
	fmt.Println("Step 3: Waiting for session to pause...")
	time.Sleep(2 * time.Second)

	// Check session status after pause
	getResult, err = client.GetSession(sessionId)
	if err != nil {
		t.Fatalf("Failed to get session: %v", err)
	}
	if !getResult.Success {
		t.Fatalf("Get session failed: %s", getResult.ErrorMessage)
	}
	if getResult.Data == nil {
		t.Fatal("Get session returned nil data")
	}
	// Session should be PAUSED or still PAUSING
	if getResult.Data.Status != "PAUSED" && getResult.Data.Status != "PAUSING" {
		t.Fatalf("Expected session status PAUSED or PAUSING, got %s", getResult.Data.Status)
	}
	t.Logf("✓ Session status after pause: %s", getResult.Data.Status)

	// Resume the session (synchronous)
	fmt.Println("Step 4: Resuming session...")
	resumeResult, err := client.Resume(createdSession, 120, 3.0)
	if err != nil {
		t.Fatalf("Failed to resume session: %v", err)
	}
	if resumeResult == nil {
		t.Fatal("Resume returned nil result")
	}
	if !resumeResult.Success {
		t.Fatalf("Resume failed: %s", resumeResult.ErrorMessage)
	}
	if resumeResult.Status != "RUNNING" {
		t.Fatalf("Expected resume status RUNNING, got %s", resumeResult.Status)
	}
	t.Logf("✓ Session resumed successfully")
	t.Logf("  Final status: %s", resumeResult.Status)
	t.Logf("  Request ID: %s", resumeResult.RequestID)

	// Verify session is RUNNING after resume
	getResult, err = client.GetSession(sessionId)
	if err != nil {
		t.Fatalf("Failed to get session: %v", err)
	}
	if !getResult.Success {
		t.Fatalf("Get session failed: %s", getResult.ErrorMessage)
	}
	if getResult.Data == nil {
		t.Fatal("Get session returned nil data")
	}
	if getResult.Data.Status != "RUNNING" {
		t.Fatalf("Expected session status RUNNING, got %s", getResult.Data.Status)
	}
	t.Logf("✓ Session is RUNNING after resume: %s", sessionId)
}

// TestSessionResumeIntegration tests successful resume operation on a session
func TestSessionResumeIntegration(t *testing.T) {
	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Fatal("AGENTBAY_API_KEY environment variable is not set")
	}

	client, err := agentbay.NewAgentBay(apiKey, nil)
	if err != nil {
		t.Fatalf("Failed to initialize AgentBay client: %v", err)
	}

	fmt.Println("Creating a new session for async resume testing...")
	createResult, err := client.Create(nil)
	if err != nil {
		t.Fatalf("Failed to create session: %v", err)
	}
	createdSession := createResult.Session
	sessionId := createdSession.SessionID
	t.Logf("Session created with ID: %s", sessionId)

	// Ensure cleanup
	defer func() {
		fmt.Println("Cleaning up: Deleting the session...")
		deleteResult, err := createdSession.Delete()
		if err != nil {
			t.Logf("Failed to delete session: %v", err)
		} else {
			t.Logf("Session %s deleted successfully", sessionId)
			if !deleteResult.Success {
				t.Logf("Session deletion failed")
			}
		}
	}()

	// Pause the session first
	fmt.Println("Step 1: Pausing session...")
	pauseResult, err := client.Pause(createdSession, 600, 2.0)
	if err != nil {
		t.Fatalf("Failed to pause session: %v", err)
	}
	if !pauseResult.Success {
		t.Fatalf("Pause failed: %s", pauseResult.ErrorMessage)
	}
	t.Logf("✓ Session pause initiated successfully")

	// Wait for pause to complete
	fmt.Println("Step 2: Waiting for session to pause...")
	time.Sleep(2 * time.Second)

	// Verify session is PAUSED or PAUSING
	getResult, err := client.GetSession(sessionId)
	if err != nil {
		t.Fatalf("Failed to get session: %v", err)
	}
	if !getResult.Success {
		t.Fatalf("Get session failed: %s", getResult.ErrorMessage)
	}
	if getResult.Data == nil {
		t.Fatal("Get session returned nil data")
	}
	if getResult.Data.Status != "PAUSED" && getResult.Data.Status != "PAUSING" {
		t.Fatalf("Expected session status PAUSED or PAUSING, got %s", getResult.Data.Status)
	}
	t.Logf("✓ Session status: %s", getResult.Data.Status)

	// Resume the session (synchronous)
	fmt.Println("Step 3: Resuming session...")
	resumeResult, err := createdSession.Resume(600, 2.0)
	if err != nil {
		t.Fatalf("Failed to async resume session: %v", err)
	}
	if resumeResult == nil {
		t.Fatal("Async resume returned nil result")
	}
	if !resumeResult.Success {
		t.Fatalf("Async resume failed: %s", resumeResult.ErrorMessage)
	}
	t.Logf("✓ Session resume completed successfully")
	t.Logf("  Request ID: %s", resumeResult.RequestID)

	// Wait a bit for resume to complete
	fmt.Println("Step 4: Waiting for session to resume...")
	time.Sleep(2 * time.Second)

	// Check session status after async resume
	getResult, err = client.GetSession(sessionId)
	if err != nil {
		t.Fatalf("Failed to get session: %v", err)
	}
	if !getResult.Success {
		t.Fatalf("Get session failed: %s", getResult.ErrorMessage)
	}
	if getResult.Data == nil {
		t.Fatal("Get session returned nil data")
	}
	// Session should be RUNNING
	if getResult.Data.Status != "RUNNING" {
		t.Fatalf("Expected session status RUNNING, got %s", getResult.Data.Status)
	}
	t.Logf("✓ Session status after resume: %s", getResult.Data.Status)
}

// TestSessionPauseNonExistent tests pause operation on non-existent session
func TestSessionPauseNonExistent(t *testing.T) {
	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Fatal("AGENTBAY_API_KEY environment variable is not set")
	}

	client, err := agentbay.NewAgentBay(apiKey, nil)
	if err != nil {
		t.Fatalf("Failed to initialize AgentBay client: %v", err)
	}

	fmt.Println("Testing pause operation on non-existent session...")
	// Create a mock session object with invalid session ID
	invalidSession := agentbay.NewSession(client, "non-existent-session-12345")

	// This should return a failed SessionPauseResult
	pauseResult, err := client.Pause(invalidSession, 600, 2.0)
	if err != nil {
		t.Fatalf("Unexpected error: %v", err)
	}
	if pauseResult == nil {
		t.Fatal("Pause returned nil result")
	}
	if pauseResult.Success {
		t.Fatal("Expected pause to fail for non-existent session")
	}
	t.Logf("✓ Returned SessionPauseResult with success=false as expected")
	t.Logf("  Error: %s", pauseResult.ErrorMessage)
	if pauseResult.ErrorMessage == "" {
		t.Error("Expected error message for non-existent session pause")
	}
}

// TestSessionResumeNonExistent tests resume operation on non-existent session
func TestSessionResumeNonExistent(t *testing.T) {
	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Fatal("AGENTBAY_API_KEY environment variable is not set")
	}

	client, err := agentbay.NewAgentBay(apiKey, nil)
	if err != nil {
		t.Fatalf("Failed to initialize AgentBay client: %v", err)
	}

	fmt.Println("Testing resume operation on non-existent session...")
	// Create a mock session object with invalid session ID
	invalidSession := agentbay.NewSession(client, "non-existent-session-12345")

	// This should return a failed SessionResumeResult
	resumeResult, err := client.Resume(invalidSession, 600, 2.0)
	if err != nil {
		t.Fatalf("Unexpected error: %v", err)
	}
	if resumeResult == nil {
		t.Fatal("Resume returned nil result")
	}
	if resumeResult.Success {
		t.Fatal("Expected resume to fail for non-existent session")
	}
	t.Logf("✓ Returned SessionResumeResult with success=false as expected")
	t.Logf("  Error: %s", resumeResult.ErrorMessage)
	if resumeResult.ErrorMessage == "" {
		t.Error("Expected error message for non-existent session resume")
	}
}

// TestSessionPauseAlreadyPaused tests pausing a session that is already paused
func TestSessionPauseAlreadyPaused(t *testing.T) {
	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Fatal("AGENTBAY_API_KEY environment variable is not set")
	}

	client, err := agentbay.NewAgentBay(apiKey, nil)
	if err != nil {
		t.Fatalf("Failed to initialize AgentBay client: %v", err)
	}

	fmt.Println("Creating a new session for already paused test...")
	createResult, err := client.Create(nil)
	if err != nil {
		t.Fatalf("Failed to create session: %v", err)
	}
	createdSession := createResult.Session
	sessionId := createdSession.SessionID
	t.Logf("Session created with ID: %s", sessionId)

	// Ensure cleanup
	defer func() {
		fmt.Println("Cleaning up: Deleting the session...")
		deleteResult, err := createdSession.Delete()
		if err != nil {
			t.Logf("Failed to delete session: %v", err)
		} else {
			t.Logf("Session %s deleted successfully", sessionId)
			if !deleteResult.Success {
				t.Logf("Session deletion failed")
			}
		}
	}()

	// Pause the session
	fmt.Println("Step 1: Pausing session...")
	pauseResult1, err := client.Pause(createdSession, 600, 2.0)
	if err != nil {
		t.Fatalf("Failed to pause session: %v", err)
	}
	if !pauseResult1.Success {
		t.Fatalf("First pause failed: %s", pauseResult1.ErrorMessage)
	}
	t.Logf("✓ Session paused successfully")

	// Wait for pause to complete
	fmt.Println("Step 2: Waiting for session to pause...")
	time.Sleep(2 * time.Second)

	// Verify session is PAUSED or PAUSING
	getResult, err := client.GetSession(sessionId)
	if err != nil {
		t.Fatalf("Failed to get session: %v", err)
	}
	if !getResult.Success {
		t.Fatalf("Get session failed: %s", getResult.ErrorMessage)
	}
	if getResult.Data == nil {
		t.Fatal("Get session returned nil data")
	}
	if getResult.Data.Status != "PAUSED" && getResult.Data.Status != "PAUSING" {
		t.Fatalf("Expected session status PAUSED or PAUSING, got %s", getResult.Data.Status)
	}
	t.Logf("✓ Session status: %s", getResult.Data.Status)

	// Try to pause again
	fmt.Println("Step 3: Attempting to pause already paused session...")
	pauseResult2, err := client.Pause(createdSession, 600, 2.0)
	if err != nil {
		t.Fatalf("Unexpected error on second pause: %v", err)
	}
	// Second pause may succeed or fail depending on API behavior
	// Both are acceptable for this test
	t.Logf("✓ Second pause completed")
	t.Logf("  Success: %t", pauseResult2.Success)
	if pauseResult2.Success {
		t.Logf("  Request ID: %s", pauseResult2.RequestID)
	} else {
		t.Logf("  Error: %s", pauseResult2.ErrorMessage)
	}
}

// TestSessionResumeAlreadyRunning tests resuming a session that is already running
func TestSessionResumeAlreadyRunning(t *testing.T) {
	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Fatal("AGENTBAY_API_KEY environment variable is not set")
	}

	client, err := agentbay.NewAgentBay(apiKey, nil)
	if err != nil {
		t.Fatalf("Failed to initialize AgentBay client: %v", err)
	}

	fmt.Println("Creating a new session for already running test...")
	createResult, err := client.Create(nil)
	if err != nil {
		t.Fatalf("Failed to create session: %v", err)
	}
	createdSession := createResult.Session
	sessionId := createdSession.SessionID
	t.Logf("Session created with ID: %s", sessionId)

	// Ensure cleanup
	defer func() {
		fmt.Println("Cleaning up: Deleting the session...")
		deleteResult, err := createdSession.Delete()
		if err != nil {
			t.Logf("Failed to delete session: %v", err)
		} else {
			t.Logf("Session %s deleted successfully", sessionId)
			if !deleteResult.Success {
				t.Logf("Session deletion failed")
			}
		}
	}()

	// Verify session is RUNNING
	getResult, err := client.GetSession(sessionId)
	if err != nil {
		t.Fatalf("Failed to get session: %v", err)
	}
	if !getResult.Success {
		t.Fatalf("Get session failed: %s", getResult.ErrorMessage)
	}
	if getResult.Data == nil {
		t.Fatal("Get session returned nil data")
	}
	if getResult.Data.Status != "RUNNING" {
		t.Fatalf("Expected session status RUNNING, got %s", getResult.Data.Status)
	}
	t.Logf("✓ Session is RUNNING: %s", sessionId)

	// Try to resume the already running session
	fmt.Println("Attempting to resume already running session...")
	resumeResult, err := client.Resume(createdSession, 30, 2)
	if err != nil {
		t.Fatalf("Unexpected error on resume: %v", err)
	}
	// Resume of already running session may succeed or have specific behavior
	// Just verify it doesn't crash and returns a proper result object
	if resumeResult == nil {
		t.Fatal("Resume returned nil result")
	}
	t.Logf("✓ Resume completed")
	t.Logf("  Success: %t", resumeResult.Success)
	if resumeResult.Success {
		t.Logf("  Status: %s", resumeResult.Status)
		t.Logf("  Request ID: %s", resumeResult.RequestID)
	} else {
		t.Logf("  Error: %s", resumeResult.ErrorMessage)
	}
}

// TestSessionPauseResumeWithCustomParameters tests pause and resume operations with custom timeout and poll interval
func TestSessionPauseResumeWithCustomParameters(t *testing.T) {
	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Fatal("AGENTBAY_API_KEY environment variable is not set")
	}

	client, err := agentbay.NewAgentBay(apiKey, nil)
	if err != nil {
		t.Fatalf("Failed to initialize AgentBay client: %v", err)
	}

	fmt.Println("Creating a new session for custom parameters test...")
	createResult, err := client.Create(nil)
	if err != nil {
		t.Fatalf("Failed to create session: %v", err)
	}
	createdSession := createResult.Session
	sessionId := createdSession.SessionID
	t.Logf("Session created with ID: %s", sessionId)

	// Ensure cleanup
	defer func() {
		fmt.Println("Cleaning up: Deleting the session...")
		deleteResult, err := createdSession.Delete()
		if err != nil {
			t.Logf("Failed to delete session: %v", err)
		} else {
			t.Logf("Session %s deleted successfully", sessionId)
			if !deleteResult.Success {
				t.Logf("Session deletion failed")
			}
		}
	}()

	// Pause with custom parameters (using agent_bay method)
	fmt.Println("Step 1: Pausing session with custom parameters...")
	pauseResult, err := client.Pause(createdSession, 5, 0.5)
	if err != nil {
		t.Fatalf("Failed to pause session with custom params: %v", err)
	}
	if pauseResult == nil {
		t.Fatal("Pause with custom params returned nil result")
	}
	if !pauseResult.Success {
		t.Fatalf("Pause with custom params failed: %s", pauseResult.ErrorMessage)
	}
	t.Logf("✓ Session pause with custom parameters successful")

	// Wait for pause to complete
	fmt.Println("Step 2: Waiting for session to pause...")
	time.Sleep(2 * time.Second)

	// Verify session is PAUSED or PAUSING
	getResult, err := client.GetSession(sessionId)
	if err != nil {
		t.Fatalf("Failed to get session: %v", err)
	}
	if !getResult.Success {
		t.Fatalf("Get session failed: %s", getResult.ErrorMessage)
	}
	if getResult.Data == nil {
		t.Fatal("Get session returned nil data")
	}
	if getResult.Data.Status != "PAUSED" && getResult.Data.Status != "PAUSING" {
		t.Fatalf("Expected session status PAUSED or PAUSING, got %s", getResult.Data.Status)
	}
	t.Logf("✓ Session status: %s", getResult.Data.Status)

	// Resume with custom parameters
	fmt.Println("Step 3: Resuming session with custom parameters...")
	resumeResult, err := client.Resume(createdSession, 300, 3.0)
	if err != nil {
		t.Fatalf("Failed to resume session with custom params: %v", err)
	}
	if resumeResult == nil {
		t.Fatal("Resume with custom params returned nil result")
	}
	if !resumeResult.Success {
		t.Fatalf("Resume with custom params failed: %s", resumeResult.ErrorMessage)
	}
	t.Logf("✓ Session resume with custom parameters successful")

	// Verify session is RUNNING after resume
	getResult, err = client.GetSession(sessionId)
	if err != nil {
		t.Fatalf("Failed to get session: %v", err)
	}
	if !getResult.Success {
		t.Fatalf("Get session failed: %s", getResult.ErrorMessage)
	}
	if getResult.Data == nil {
		t.Fatal("Get session returned nil data")
	}
	if getResult.Data.Status != "RUNNING" {
		t.Fatalf("Expected session status RUNNING, got %s", getResult.Data.Status)
	}
	t.Logf("✓ Session is RUNNING after resume with custom parameters")
}

// TestSessionPauseWithShortTimeout tests pause with a short timeout parameter
func TestSessionPauseWithShortTimeout(t *testing.T) {
	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Fatal("AGENTBAY_API_KEY environment variable is not set")
	}

	client, err := agentbay.NewAgentBay(apiKey, nil)
	if err != nil {
		t.Fatalf("Failed to initialize AgentBay client: %v", err)
	}

	fmt.Println("Creating a new session for short timeout test...")
	createResult, err := client.Create(nil)
	if err != nil {
		t.Fatalf("Failed to create session: %v", err)
	}
	createdSession := createResult.Session
	sessionId := createdSession.SessionID
	t.Logf("Session created with ID: %s", sessionId)

	// Ensure cleanup
	defer func() {
		fmt.Println("Cleaning up: Deleting the session...")
		deleteResult, err := createdSession.Delete()
		if err != nil {
			t.Logf("Failed to delete session: %v", err)
		} else {
			t.Logf("Session %s deleted successfully", sessionId)
			if !deleteResult.Success {
				t.Logf("Session deletion failed")
			}
		}
	}()

	// Pause with short timeout (using agent_bay method)
	fmt.Println("Pausing session with short timeout...")
	pauseResult, err := client.Pause(createdSession, 10, 1.0)
	if err != nil {
		t.Fatalf("Failed to pause session with short timeout: %v", err)
	}
	if pauseResult == nil {
		t.Fatal("Pause with short timeout returned nil result")
	}
	// Pause should succeed regardless of timeout parameter
	if !pauseResult.Success {
		t.Fatalf("Pause with short timeout failed: %s", pauseResult.ErrorMessage)
	}
	t.Logf("✓ Pause with short timeout successful")
}

File: java/agentbay/src/integration-test/java/com/aliyun/agentbay/test/SessionPauseResumeIntegrationTest.java
// ci-stable
package com.aliyun.agentbay.test;

import com.aliyun.agentbay.AgentBay;
import com.aliyun.agentbay.exception.AgentBayException;
import com.aliyun.agentbay.model.*;
import com.aliyun.agentbay.model.SessionListResult.SessionInfo;
import com.aliyun.agentbay.session.CreateSessionParams;
import com.aliyun.agentbay.session.Session;
import org.junit.After;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import static org.junit.Assert.*;

/**
 * Integration tests for Session pause and resume operations.
 * 
 * This test class is equivalent to TestSessionPauseResumeIntegration in 
 * test_agentbay_list_status.py
 */
public class SessionPauseResumeIntegrationTest {
    private static AgentBay agentBay;
    private List<Session> testSessions;

    @BeforeClass
    public static void setUpClass() throws AgentBayException {
        // Initialize AgentBay client
        String separator = "============================================================";
        System.out.println("\n" + separator);
        System.out.println("ENVIRONMENT CONFIGURATION");
        System.out.println(separator);
        
        String apiKey = System.getenv("AGENTBAY_API_KEY");
        if (apiKey == null || apiKey.isEmpty()) {
            System.out.println("API Key: ✗ Not Set");
            throw new RuntimeException("AGENTBAY_API_KEY environment variable not set");
        }
        System.out.println("API Key: ✓ Set");
        
        String endpoint = System.getenv("AGENTBAY_ENDPOINT");
        if (endpoint != null && !endpoint.isEmpty()) {
            System.out.println("Endpoint: " + endpoint);
            com.aliyun.agentbay.Config config = new com.aliyun.agentbay.Config();
            config.setEndpoint(endpoint);
            agentBay = new AgentBay(apiKey, config);
        } else {
            System.out.println("Endpoint: Using default");
            agentBay = new AgentBay(apiKey);
        }
        System.out.println(separator + "\n");
    }

    @Before
    public void setUp() {
        // Initialize test sessions list for cleanup
        testSessions = new ArrayList<>();
    }

    @After
    public void tearDown() {
        // Clean up test sessions after each test
        System.out.println("\nCleaning up test sessions for this test...");
        for (Session session : testSessions) {
            try {
                if (session != null) {
                    // Get session status first
                    SessionStatusResult statusResult = session.getStatus();
                    System.out.println("  ✓ Session status: " + session.getSessionId() + " - " + statusResult.getStatus());
                    
                    // TODO: Implement betaResume method
                    // If session is PAUSED, resume it first
                    if ("PAUSED".equals(statusResult.getStatus())) {
                        SessionResumeResult resumeResult = session.betaResume();
                        if (resumeResult.isSuccess()) {
                            System.out.println("  ✓ Resumed session: " + session.getSessionId());
                        }
                    }
                    
                    // Delete session if not already deleting/deleted
                    if (statusResult.getStatus() != null && 
                        !statusResult.getStatus().equals("DELETING") && 
                        !statusResult.getStatus().equals("DELETED") &&
                        !statusResult.getStatus().equals("RESUMING") &&
                        !statusResult.getStatus().equals("PAUSING") &&
                        !statusResult.getStatus().equals("FINISH")) {
                        com.aliyun.agentbay.model.DeleteResult deleteResult = session.delete(false);
                        if (deleteResult.isSuccess()) {
                            System.out.println("  ✓ Deleted session: " + session.getSessionId());
                        } else {
                            System.out.println("  ✗ Failed to delete session: " + session.getSessionId());
                        }
                    }
                }
            } catch (Exception e) {
                System.out.println("  ✗ Error cleaning up session: " + e.getMessage());
            }
        }
        testSessions.clear();
    }

    /**
     * Helper method to create a test session.
     */
    private Session createTestSession() throws AgentBayException {
        String sessionName = "test-pause-resume-" + UUID.randomUUID().toString().substring(0, 8);
        System.out.println("\nCreating test session: " + sessionName);

        // Create session with labels
        Map<String, String> labels = new HashMap<>();
        labels.put("project", "piaoyun-demo");
        labels.put("environment", "testing");
        
        CreateSessionParams params = new CreateSessionParams();
        params.setLabels(labels);
        
        SessionResult result = agentBay.create(params);
        assertTrue("Failed to create session: " + result.getErrorMessage(), result.isSuccess());
        assertNotNull("Session is null", result.getSession());

        Session session = result.getSession();
        testSessions.add(session);
        System.out.println("  ✓ Session created: " + session.getSessionId());

        return session;
    }

    /**
     * Helper method to verify session status using both get_status and _get_session,
     * and verify the session appears in the list with correct status.
     */
    private String verifySessionStatusAndList(Session session, String[] expectedStatuses, String operationName) 
            throws AgentBayException, InterruptedException {
        System.out.println("\nVerifying session status after " + operationName + "...");
        
        // First call getStatus to check the current status
        SessionStatusResult statusResult = session.getStatus();
        assertTrue("Failed to get session status: " + statusResult.getErrorMessage(), 
                  statusResult.isSuccess());
        
        String initialStatus = statusResult.getStatus() != null ? statusResult.getStatus() : "UNKNOWN";
        System.out.println("  ✓ Session status from getStatus: " + initialStatus);
        
        // Verify status is one of expected statuses
        boolean statusMatched = false;
        for (String expected : expectedStatuses) {
            if (expected.equals(initialStatus)) {
                statusMatched = true;
                break;
            }
        }
        assertTrue("Unexpected status " + initialStatus + ", expected one of " + 
                  String.join(", ", expectedStatuses), statusMatched);

        if ("FINISH".equals(initialStatus)) {
            return initialStatus;
        }

        // Then call _getSession for detailed information
        GetSessionResult sessionInfo = agentBay._getSession(session.getSessionId());
        assertTrue("Failed to get session info: " + sessionInfo.getErrorMessage(), 
                  sessionInfo.isSuccess());

        String currentStatus = sessionInfo.getData() != null ? 
                              sessionInfo.getData().getStatus() : "UNKNOWN";
        assertEquals("Session status mismatch: expected " + initialStatus + ", got " + currentStatus,
                    initialStatus, currentStatus);
        System.out.println("  ✓ Session status from _getSession: " + currentStatus);
        
        // TODO: Implement list method with status filter
        // Java AgentBay doesn't have a list(status) method yet
        // For now, we skip the list verification part
        // When list method is implemented, uncomment the following code:
        
        SessionListResult listResult = agentBay.list(currentStatus);
        assertTrue("Failed to list sessions: " + listResult.getErrorMessage(), 
                  listResult.isSuccess());
        
        // Verify session is in the list and check structure
        boolean sessionFound = false;
        for (SessionInfo sessionData : listResult.getSessionInfos()) {
            String sessionId = (String) sessionData.getSessionId();
            if (session.getSessionId().equals(sessionId)) {
                sessionFound = true;
                assertNotNull("sessionStatus field missing in list result", 
                          sessionData.getSessionStatus());
                assertNotNull("sessionId field missing in list result", 
                          sessionData.getSessionId());
                assertEquals("Session status mismatch in list", 
                           currentStatus, sessionData.getSessionStatus());
                break;
            }
        }
        
        assertTrue("Session " + session.getSessionId() + " not found in list with status " + currentStatus,
                  sessionFound);
        System.out.println("  ✓ Session found in list with status " + currentStatus);
        
        System.out.println("  ✓ Session status verification completed for " + operationName);
        
        return currentStatus;
    }

    @Test
    public void testPauseSessionSuccess() throws Exception {
        String separator = "============================================================";
        System.out.println("\n" + separator);
        System.out.println("TEST: Pause Session Success");
        System.out.println(separator);

        // Step 1: Create a test session
        Session session = createTestSession();

        // Step 2: Verify session is initially in RUNNING state
        SessionStatusResult statusResult = session.getStatus();
        assertTrue("Failed to get session status: " + statusResult.getErrorMessage(), 
                  statusResult.isSuccess());
        
        String initialStatus = statusResult.getStatus() != null ? statusResult.getStatus() : "UNKNOWN";
        System.out.println("  ✓ Session status from getStatus: " + initialStatus);
        assertEquals("Unexpected initial status, expected RUNNING", "RUNNING", initialStatus);

        // Step 3: Pause the session
        System.out.println("\nStep 2: Pausing session...");
        SessionPauseResult pauseResult = agentBay.betaPause(session);

        // Verify pause result
        assertNotNull("Pause result is null", pauseResult);
        assertTrue("Pause failed: " + pauseResult.getErrorMessage(), pauseResult.isSuccess());
        System.out.println("  ✓ Session pause initiated successfully");
        System.out.println("    Request ID: " + pauseResult.getRequestId());

        // Step 4: Wait for pause to complete
        System.out.println("\nStep 3: Waiting for session to pause...");
        Thread.sleep(2000);

        // Step 5: Verify session status after pause
        String currentStatus = verifySessionStatusAndList(session, 
                                                         new String[]{"PAUSED", "PAUSING"}, 
                                                         "pause");
        
        System.out.println("\n✅ Test completed successfully：" + currentStatus);
    }

    @Test
    public void testPauseAndDeleteSessionSuccess() throws Exception {
        String separator = "============================================================";
        System.out.println("\n" + separator);
        System.out.println("TEST: Pause and Delete Session Success");
        System.out.println(separator);

        // Step 1: Create a test session
        System.out.println("\nStep 1: Creating test session...");
        Session session = createTestSession();

        // Step 2: Pause the session
        System.out.println("\nStep 2: Pausing session...");
        SessionPauseResult pauseResult = agentBay.betaPause(session);

        // Verify pause result
        assertNotNull("Pause result is null", pauseResult);
        assertTrue("Pause failed: " + pauseResult.getErrorMessage(), pauseResult.isSuccess());
        System.out.println("  ✓ Session pause initiated successfully");
        System.out.println("    Request ID: " + pauseResult.getRequestId());

        // Step 3: Wait for pause to complete
        System.out.println("\nStep 3: Waiting for session to pause...");
        Thread.sleep(2000);

        // Step 4: Check session status before resuming
        System.out.println("  ✓ Checking session status before resuming");
        
        // TODO: Implement betaResume method
        // Resume the session before deletion
        SessionResumeResult resumeResult = session.betaResume();
        assertTrue("Resume failed: " + resumeResult.getErrorMessage(), resumeResult.isSuccess());
        System.out.println("  ✓ Session resumed");

        // Step 5: Delete the session
        System.out.println("\nStep 4: Deleting session...");
        com.aliyun.agentbay.model.DeleteResult deleteResult = session.delete(false);
        if (deleteResult.isSuccess()) {
            System.out.println("  ✓ Delete session successfully");
        }

        // Step 6: Verify session status after delete
        String currentStatus = verifySessionStatusAndList(session, 
                                                         new String[]{"DELETING", "DELETED", "FINISH"}, 
                                                         "delete");
        
        System.out.println("\n✅ Test completed successfully");
    }

    // TODO: Implement test_resume_async_session_success after betaResume is implemented
    @Test
    public void testResumeAsyncSessionSuccess() throws Exception {
        String separator = "============================================================";
        System.out.println("\n" + separator);
        System.out.println("TEST: Async Resume Session Success");
        System.out.println("\n" + separator);
        // Step 1: Create a test session
        Session session = createTestSession();
    
        // Step 2: Pause the session first
        System.out.println("\nStep 1: Pausing session...");
        SessionPauseResult pauseResult = agentBay.betaPause(session);
        assertTrue("Pause failed: " + pauseResult.getErrorMessage(), pauseResult.isSuccess());
        System.out.println("  ✓ Session pause initiated successfully");
    
        // Step 3: Wait for pause to complete
        System.out.println("\nStep 2: Waiting for session to pause...");
        Thread.sleep(2000);
    
        // Step 4: Verify session is PAUSED or PAUSING
        SessionStatusResult statusResult = session.getStatus();
        assertTrue("Failed to get session status: " + statusResult.getErrorMessage(), 
                  statusResult.isSuccess());
        Thread.sleep(2000);
        
        String status = statusResult.getStatus() != null ? statusResult.getStatus() : "UNKNOWN";
        System.out.println("  ✓ Session status from getStatus: " + status);
        assertTrue("Unexpected status, expected PAUSED or PAUSING", 
                  "PAUSED".equals(status) || "PAUSING".equals(status));
        System.out.println("  ✓ Session status checked");
    
        // Step 5: Resume the session (asynchronous)
        System.out.println("\nStep 3: Resuming session asynchronously...");
        SessionResumeResult resumeResult = agentBay.betaResume(session);
    
        // Verify async resume result
        assertNotNull("Resume result is null", resumeResult);
        assertTrue("Async resume failed: " + resumeResult.getErrorMessage(), 
                  resumeResult.isSuccess());
        System.out.println("  ✓ Session resume initiated successfully");
        System.out.println("    Request ID: " + resumeResult.getRequestId());
    
        // Step 6: Wait for resume to complete
        System.out.println("\nStep 4: Waiting for session to resume...");
        Thread.sleep(2000);
    
        // Step 7: Verify session status after async resume
        String currentStatus = verifySessionStatusAndList(session, 
                                                         new String[]{"RUNNING", "RESUMING"}, 
                                                         "async resume");
        
        System.out.println("\n✅ Test completed successfully");
    }
}
